File: backend/src/test/kotlin/com/aicoach/service/DashboardServiceTest.kt
package com.aicoach.service

import com.aicoach.domain.dto.WeekAdherenceDto
import com.aicoach.domain.entity.StravaActivity
import com.aicoach.domain.entity.TrainingPlan
import com.aicoach.domain.entity.WeeklyBlock
import com.aicoach.repository.StravaActivityRepository
import com.aicoach.repository.TrainingPlanRepository
import com.aicoach.repository.WeeklyBlockRepository
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import java.math.BigDecimal
import java.time.LocalDate
import java.util.Optional

class DashboardServiceTest : DescribeSpec({

    val planRepository = mockk<TrainingPlanRepository>()
    val weeklyBlockRepository = mockk<WeeklyBlockRepository>()
    val activityRepository = mockk<StravaActivityRepository>()
    val service = DashboardService(planRepository, weeklyBlockRepository, activityRepository)

    fun makePlan(id: Long = 1L) = TrainingPlan(id = id, name = "Test Plan", totalWeeks = 2)

    fun makeWeek(
        planId: Long = 1L,
        weekNumber: Int = 1,
        start: LocalDate = LocalDate.of(2025, 1, 6),
        end: LocalDate = LocalDate.of(2025, 1, 12),
        plannedKm: Double? = 50.0,
        plannedVertM: Int? = 500
    ): WeeklyBlock {
        val plan = TrainingPlan(id = planId, name = "Test Plan", totalWeeks = 2)
        return WeeklyBlock(
            id = weekNumber.toLong(),
            trainingPlan = plan,
            weekNumber = weekNumber,
            startDate = start,
            endDate = end,
            plannedKm = plannedKm,
            plannedVertM = plannedVertM
        )
    }

    fun makeActivity(
        id: Long = 1L,
        athleteId: Long = 0L,
        date: LocalDate = LocalDate.of(2025, 1, 7),
        distanceM: Double = 10000.0,
        elevationM: Double = 100.0,
        sportType: String = "Run"
    ) = StravaActivity(
        id = id,
        stravaId = id * 100,
        athleteId = athleteId,
        activityDate = date,
        distanceM = BigDecimal.valueOf(distanceM),
        totalElevationM = BigDecimal.valueOf(elevationM),
        sportType = sportType
    )

    describe("getSummary") {
        it("returns empty DTO when no plan exists") {
            every { planRepository.findFirstByOrderByIdAsc() } returns Optional.empty()

            val result = service.getSummary()

            result.weeks shouldBe emptyList()
            result.currentWeekNumber shouldBe null
            result.totalPlannedKm shouldBe 0.0
            result.totalActualKm shouldBe 0.0
        }

        it("calculates adherence percent correctly") {
            val plan = makePlan()
            val week = makeWeek(plannedKm = 50.0)
            val activities = listOf(makeActivity(distanceM = 40000.0)) // 40 km

            every { planRepository.findFirstByOrderByIdAsc() } returns Optional.of(plan)
            every { weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(1L) } returns listOf(week)
            every {
                activityRepository.findByActivityDateBetweenOrderByActivityDateAsc(any(), any())
            } returns activities

            val result = service.getSummary()

            result.weeks[0].adherencePercent shouldBe 80.0 // 40/50 * 100
            result.weeks[0].actualKm shouldBe 40.0
        }

        it("caps adherence at 200.0") {
            val plan = makePlan()
            val week = makeWeek(plannedKm = 10.0)
            val activities = listOf(makeActivity(distanceM = 25000.0)) // 25 km

            every { planRepository.findFirstByOrderByIdAsc() } returns Optional.of(plan)
            every { weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(1L) } returns listOf(week)
            every {
                activityRepository.findByActivityDateBetweenOrderByActivityDateAsc(any(), any())
            } returns activities

            val result = service.getSummary()

            result.weeks[0].adherencePercent shouldBe 200.0
        }

        it("filters out non-Run sport types") {
            val plan = makePlan()
            val week = makeWeek(plannedKm = 50.0)
            val activities = listOf(
                makeActivity(id = 1L, distanceM = 20000.0, sportType = "Run"),
                makeActivity(id = 2L, distanceM = 30000.0, sportType = "Ride"), // should be filtered
                makeActivity(id = 3L, distanceM = 10000.0, sportType = "TrailRun")
            )

            every { planRepository.findFirstByOrderByIdAsc() } returns Optional.of(plan)
            every { weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(1L) } returns listOf(week)
            every {
                activityRepository.findByActivityDateBetweenOrderByActivityDateAsc(any(), any())
            } returns activities

            val result = service.getSummary()

            result.weeks[0].actualKm shouldBe 30.0 // 20 + 10, Ride excluded
            result.weeks[0].activityCount shouldBe 2
        }

        it("correctly identifies isCurrentWeek and isFutureWeek") {
            val today = LocalDate.now()
            val plan = makePlan()
            val currentWeek = makeWeek(
                weekNumber = 1,
                start = today.minusDays(2),
                end = today.plusDays(4)
            )
            val futureWeek = makeWeek(
                weekNumber = 2,
                start = today.plusDays(7),
                end = today.plusDays(13)
            )

            every { planRepository.findFirstByOrderByIdAsc() } returns Optional.of(plan)
            every { weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(1L) } returns listOf(currentWeek, futureWeek)
            every {
                activityRepository.findByActivityDateBetweenOrderByActivityDateAsc(any(), any())
            } returns emptyList()

            val result = service.getSummary()

            result.weeks[0].isCurrentWeek shouldBe true
            result.weeks[0].isFutureWeek shouldBe false
            result.weeks[1].isCurrentWeek shouldBe false
            result.weeks[1].isFutureWeek shouldBe true
        }
    }

    describe("getSummaryForAthlete") {
        it("returns empty DTO when athlete has no plan") {
            every { planRepository.findByAthleteId(5L) } returns Optional.empty()

            val result = service.getSummaryForAthlete(5L)

            result.weeks shouldBe emptyList()
            result.currentWeekNumber shouldBe null
        }

        it("calculates adherence for specific athlete activities") {
            val plan = makePlan()
            val week = makeWeek(plannedKm = 60.0)
            val activities = listOf(
                makeActivity(id = 1L, athleteId = 5L, distanceM = 30000.0),
                makeActivity(id = 2L, athleteId = 5L, distanceM = 30000.0)
            )

            every { planRepository.findByAthleteId(5L) } returns Optional.of(plan)
            every { weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(1L) } returns listOf(week)
            every {
                activityRepository.findByInternalAthleteIdAndActivityDateBetweenOrderByActivityDateAsc(5L, any(), any())
            } returns activities

            val result = service.getSummaryForAthlete(5L)

            result.weeks[0].actualKm shouldBe 60.0
            result.weeks[0].adherencePercent shouldBe 100.0
        }
    }
})

File: server/src/services/dashboardService.test.ts
import { describe, it, expect, beforeAll, afterEach } from 'vitest'
import { createTestDb, resetDb } from '../test/dbHelper.js'
import { run, queryOne } from '../db/client.js'
import { getDashboardSummaryForAthlete } from './dashboardService.js'

beforeAll(async () => {
  await createTestDb()
})

afterEach(() => {
  resetDb()
})

function insertAthlete(name = 'Test Athlete'): number {
  const now = new Date().toISOString()
  run(
    `INSERT INTO athletes (name, trail_access, strava_enabled, created_at, updated_at) VALUES (?, 0, 0, ?, ?)`,
    [name, now, now]
  )
  return queryOne('SELECT last_insert_rowid() as id')!.id as number
}

function insertPlan(athleteId: number, totalWeeks = 4): number {
  const now = new Date().toISOString()
  run(
    `INSERT INTO training_plans (athlete_id, name, total_weeks, created_at) VALUES (?, ?, ?, ?)`,
    [athleteId, 'Test Plan', totalWeeks, now]
  )
  return queryOne('SELECT last_insert_rowid() as id')!.id as number
}

function insertWeek(
  planId: number,
  weekNumber: number,
  startDate: string,
  endDate: string,
  plannedKm = 50,
  plannedVertM = 500,
  phase = 'Base'
): number {
  run(
    `INSERT INTO weekly_blocks (training_plan_id, week_number, phase, start_date, end_date, planned_km, planned_vert_m)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [planId, weekNumber, phase, startDate, endDate, plannedKm, plannedVertM]
  )
  return queryOne('SELECT last_insert_rowid() as id')!.id as number
}

describe('getDashboardSummaryForAthlete', () => {
  it('returns empty weeks array when athlete has no plan', () => {
    const athleteId = insertAthlete()
    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result).toEqual({
      weeks: [],
      currentWeekNumber: null,
      totalPlannedKm: 0,
      totalActualKm: 0,
    })
  })

  it('returns empty for non-existent athlete', () => {
    const result = getDashboardSummaryForAthlete(99999)
    expect(result).toEqual({
      weeks: [],
      currentWeekNumber: null,
      totalPlannedKm: 0,
      totalActualKm: 0,
    })
  })

  it('isCurrentWeek=true for week spanning today', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    const start = new Date(Date.now() - 3 * 86400000).toISOString().substring(0, 10)
    const end = new Date(Date.now() + 3 * 86400000).toISOString().substring(0, 10)
    insertWeek(planId, 1, start, end)

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.weeks).toHaveLength(1)
    expect(result.weeks[0].isCurrentWeek).toBe(true)
    expect(result.weeks[0].isFutureWeek).toBe(false)
    expect(result.currentWeekNumber).toBe(1)
  })

  it('isCurrentWeek=false and isFutureWeek=false for past week', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07')

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.weeks[0].isCurrentWeek).toBe(false)
    expect(result.weeks[0].isFutureWeek).toBe(false)
    expect(result.currentWeekNumber).toBeNull()
  })

  it('isFutureWeek=true for week in the future', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    insertWeek(planId, 1, '2099-01-01', '2099-01-07')

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.weeks[0].isFutureWeek).toBe(true)
    expect(result.weeks[0].isCurrentWeek).toBe(false)
  })

  it('totalPlannedKm excludes future weeks', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId, 2)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07', 50)
    insertWeek(planId, 2, '2099-01-01', '2099-01-07', 60)

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.totalPlannedKm).toBe(50)
  })

  it('currentWeekNumber is null when no week spans today', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId, 2)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07')
    insertWeek(planId, 2, '2099-01-01', '2099-01-07')

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.currentWeekNumber).toBeNull()
  })

  it('adherencePercent is 0 when plannedKm is 0 (no divide-by-zero)', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07', 0)

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.weeks[0].adherencePercent).toBe(0)
  })

  it('actualKm is 0.00 when no activities', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07')

    const result = getDashboardSummaryForAthlete(athleteId)
    expect(result.weeks[0].actualKm).toBe(0)
    expect(result.weeks[0].actualVertM).toBe(0)
  })

  it('week fields include phase and date strings', () => {
    const athleteId = insertAthlete()
    const planId = insertPlan(athleteId)
    insertWeek(planId, 1, '2020-01-01', '2020-01-07', 50, 500, 'Build')

    const result = getDashboardSummaryForAthlete(athleteId)
    const week = result.weeks[0]
    expect(week.weekNumber).toBe(1)
    expect(week.phase).toBe('Build')
    expect(week.startDate).toBe('2020-01-01')
    expect(week.endDate).toBe('2020-01-07')
    expect(week.plannedKm).toBe(50)
    expect(week.plannedVertM).toBe(500)
  })
})
